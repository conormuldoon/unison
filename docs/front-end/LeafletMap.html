<!DOCTYPE html>
<html lang="en">
<head>
    
    <meta charset="utf-8">
    <title>JSDoc: LeafletMap</title>

    <script src="https://cdn.jsdelivr.net/gh/google/code-prettify@master/loader/run_prettify.js"></script>
    <script src="https://unpkg.com/@babel/standalone/babel.min.js"></script>
    <script src="./build/entry.js"></script>
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link href="https://fonts.googleapis.com/css?family=Muli:100,400,700|Oswald:300|Inconsolata,700" rel="stylesheet">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.6.3/css/all.css" integrity="sha384-UHRtZLI+pbxtHCWp1t77Bi1L4ZtiqrqD80Kn4Z8NTSRyMA2Fd33n5dQ8lWUE00s/" crossorigin="anonymous">
    <link type="text/css" rel="stylesheet" href="https://jmblog.github.io/color-themes-for-google-code-prettify/themes/tomorrow-night.min.css">
    <link type="text/css" rel="stylesheet" href="styles/app.min.css">
    <link type="text/css" rel="stylesheet" href="styles/iframe.css">
</head>

<body>
    <div id="stickyNavbarOverlay"></div>
    <div class="top-navbar">
        <div class="container">
            <nav class="navbar" role="navigation" aria-label="main navigation">
                <div class="navbar-brand">
                    
                     
                        <h1 class="navbar-item">Unison React components</h1>
                    
                    <a id="hamburger" role="button" class="navbar-burger" aria-label="menu" aria-expanded="false">
                        <span aria-hidden="true"></span>
                        <span aria-hidden="true"></span>
                        <span aria-hidden="true"></span>
                    </a>
                </div>
                
            </nav>
        </div>
    </div>
    <div class="container">
        <div class="columns">
            <div class="column is-3" id="sidebarNav">
                <div class="sidebar">
                    <nav>
                        <h2><a href="index.html">Home</a></h2><div class="category"><h3>Components</h3><ul><li><a href="ARLocationComponent.html">ARLocationComponent</a></li><li><a href="ChartComponent.html">ChartComponent</a></li><li><a href="ChartPopup.html">ChartPopup</a></li><li><a href="DateSelector.html">DateSelector</a></li><li><a href="LeafletMap.html">LeafletMap</a></li><li><a href="LocationForm.html">LocationForm</a></li><li><a href="RemoveComponent.html">RemoveComponent</a></li><li><a href="TabsComponent.html">TabsComponent</a></li><li><a href="Unison.html">Unison</a></li></ul></div>
                    </nav>
                </div>
            </div>
            <div class="column is-9-desktop">
                <div class="content" id="main-content-wrapper">
                    <header class="page-title">
                        <p>Components</p>
                        <h1>LeafletMap</h1>
                    </header>
                    
                    




<section>

<header>
    
        <h2>
            &ltLeafletMap <i>curVar</i> <i>fromDate</i> <i>toDate</i> <i>marker</i> <i>markerCallback</i> <i>mapCentre</i> <i>featureProperties</i> <i>linksProperty</i> /&gt
        </h2>
        
            <div class="class-description"><p>A component for displaying a Leaflet map and markers for popups for locations where weather data is being tracked.</p></div>
        
    
</header>

<article>
    <div class="container-overview">
    
        
            <div class='vertical-section'>
                <h3 class="title"><span>Constructor</span></h3>
                <div class="members">
                    <div class="member">


    
    <h4 class="name" id="LeafletMap">
        <a class="href-link" href="#LeafletMap">#</a>
        
        <span class="code-name">
            
                &ltLeafletMap <i>curVar</i> <i>fromDate</i> <i>toDate</i> <i>marker</i> <i>markerCallback</i> <i>mapCentre</i> <i>featureProperties</i> <i>linksProperty</i> /&gt
            
        </span>
    </h4>
    

    
    







    
        <h5>PropTypes:</h5>
        

<table class="params">
  <thead>
  <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Required</th>
      <th>Description</th>
      <th class="last">Default</th>
  </tr>
  </thead>

  <tbody>
  
        <tr class="deep-level-0">
          <td class="name"><code>curVar</code></td>
          <td>
            string
          </td>
          <td>No</td>
          <td>Specifies the weather variable currently selected.</td>
          <td></td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>fromDate</code></td>
          <td>
            string
          </td>
          <td>No</td>
          <td>Specifies the start date for the data that is to be displayed.</td>
          <td></td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>toDate</code></td>
          <td>
            string
          </td>
          <td>No</td>
          <td>Specifies the end date for the data that is to be displayed.</td>
          <td></td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>marker</code></td>
          <td>
            array
          </td>
          <td>No</td>
          <td>An array of markers to be displayed on the map. Each element of the array specifies the coordinates and location 
name associated with a marker.</td>
          <td></td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>markerCallback</code></td>
          <td>
            func
          </td>
          <td>No</td>
          <td>A function called when a marker is clicked.</td>
          <td></td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>mapCentre</code></td>
          <td>
            array
          </td>
          <td>Yes</td>
          <td>The latitude/longitude coordinates for the centre of the map.</td>
          <td></td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>featureProperties</code></td>
          <td>
            object
          </td>
          <td>No</td>
          <td></td>
          <td></td>
        </tr>
        
  
        <tr class="deep-level-0">
          <td class="name"><code>linksProperty</code></td>
          <td>
            object
          </td>
          <td>No</td>
          <td></td>
          <td></td>
        </tr>
        
  
  </tbody>
</table>
    







<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    
    

    

    

    
        <p class="tag-source">
            <a href="LeafletMap.js.html" class="button">View Source</a>
            <span>
                <a href="LeafletMap.js.html">LeafletMap.js</a>, <a href="LeafletMap.js.html#line24">line 24</a>
            </span>
        </p>
    
</dl>





















</div>
                </div>
            </div>
        
    
    </div>
    
        
    
    

    

    

    

    

    

    

    

    

    
</article>

</section>




                </div>
            </div>
        </div>
    </div>


<footer class="footer">
    <div class="content has-text-centered">
        <p>Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.6.3</a> on Thu May 28 2020 20:21:36 GMT+0100 (Irish Standard Time)</p>
        <p class="sidebar-created-by">
            <a href="https://github.com/SoftwareBrothers/better-docs" target="_blank">BetterDocs theme</a> provided with <i class="fas fa-heart"></i> by 
            <a href="http://softwarebrothers.co" target="_blank">SoftwareBrothers - JavaScript Development Agency</a>
        </p>
    </div>
</footer>


<script src="scripts/app.min.js"></script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>