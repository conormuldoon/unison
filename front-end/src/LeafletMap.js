
import React, { Component  } from 'react';

import { Map, Marker, TileLayer} from 'react-leaflet';
import Leaflet from 'leaflet';
import ChartPopup from './ChartPopup';

import PropTypes from 'prop-types';

const image = new Leaflet.Icon({
               iconUrl: require('./2000px-Map_marker.png'),
               iconSize:     [30, 46],
           })


/**
 * A component for displaying a Leaflet map and markers for popups for locations where weather data is being tracked.
 * 
 * @component
 * 
 */

 class LeafletMap extends Component {
  constructor(props) {
    super(props)
    this.state = {

      zoom: 12,
      popupComponent: undefined,
      dragging: true,

    }
  }

  componentDidUpdate = (prevProps, prevState) =>{

    if(prevProps.curVar!==this.props.curVar||prevProps.curLoc!==this.props.curLoc||prevProps.fromDate!==this.props.fromDate||prevProps.toDate!==this.props.toDate){
      if(this.state.popupComponent!==undefined){
        this.closePopup();
        this.addPopup(this.props.curLoc);
      }
    }
  }

  addPopup = (name) =>{
    this.props.markerCallback(name);

    const popupComponent=<ChartPopup varCur={this.props.curVar} location={name} fromDate={this.props.fromDate} toDate={this.props.toDate} closePopup={this.closePopup}/>;
    this.setState({popupComponent:popupComponent,dragging:false});


  }

  markerCallback = (marker) => {

    if(this.state.popupComponent!==undefined){
        this.closePopup();
    }

    this.addPopup(marker.name);

  }

  closePopup = (e) =>{

    this.setState({popupComponent:undefined,dragging:true});

  }



  render() {
    

    return (
      <Map center={this.props.mapCentre} zoom={this.state.zoom} dragging={this.state.dragging} >
        <TileLayer
          attribution='&copy; <a href="https://osm.org/copyright">OpenStreetMap</a> contributors'
          url='https://{s}.tile.osm.org/{z}/{x}/{y}.png'
        />

        {this.props.marker&&this.props.marker.map((marker) =>
         <Marker key={marker.name} position={marker.position} onClick={this.markerCallback.bind(this,marker)} icon={image}/>

       )}
       <div id="marginclickdiv" onClick={this.closePopup}>
       {this.state.popupComponent}
       </div>
      </Map>
    );
  }
}

LeafletMap.propTypes ={
  /** Specifies the weather variable currently selected. */
  curVar: PropTypes.string,

  /** Sepecifies the location selected. */
  curLoc: PropTypes.string,

  /** Specifies the start date for the data that is to be displayed. */
  fromDate: PropTypes.string,

  /** Specifies the end date for the data that is to be displayed. */
  toDate: PropTypes.string,

  /** An array of markers to be displayed on the map. Each element of the array specifies the coordinates and location 
   * name associated with a marker.*/
  marker: PropTypes.array,

  /** A function called when a marker is clicked. */
  markerCallback: PropTypes.func,

  /** The latitude/longitude coordinates for the centre of the map. */
  mapCentre: PropTypes.array.isRequired,

}

export default LeafletMap;
