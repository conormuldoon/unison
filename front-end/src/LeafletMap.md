LeafletMap example:

```js
<div id="mapdiv">
    <LeafletMap curVar="Precipitation" curLoc="Oslo" fromDate="4/10/2019" toDate="5/10/2019" />
</div>
```