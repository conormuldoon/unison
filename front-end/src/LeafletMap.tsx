
import Leaflet from 'leaflet';
import PropTypes from 'prop-types';
import React, { useState, useEffect } from 'react';
import { Map, TileLayer } from 'react-leaflet';



let uri = require('./2000px-Map_marker.png');
if (uri.default) {
  uri = uri.default;
}

const image = new Leaflet.Icon({
  iconUrl: uri,
  iconSize: [30, 46],
})


export type MapMarkerFactory = (component: JSX.Element,
  callback: (name: string) => void,
  image: Leaflet.Icon)
  => JSX.Element;


export const DEFAULT_ZOOM = 12;

export type PopupFactory = ((closePopup: () => void) => React.ReactNode);

export type MapFactory = (marker: MapMarkerFactory[],
  markerClicked: (location: string) => void,
  popupFactory?: PopupFactory) => JSX.Element;


export function createMapFactory(mapCentre: [number, number]): MapFactory {

  return function mapFactory(marker, markerClicked, popupFactory) {

    return <LeafletMap marker={marker}
      markerCallback={markerClicked}
      mapCentre={mapCentre} popupFactory={popupFactory} />;

  }

}

export interface MapProps {

  /**
   * A callback invoked when a marker on the map is clicked.
   */
  markerCallback: (location: string) => void;

  /**
   * A tuple of two numbers for the map's centre.
   */
  mapCentre: [number, number];

  /**
   * An array of markers that are displayed on the map.
   */
  marker: MapMarkerFactory[];

  /**
   * An optional factory that creates chart popup components.
   */
  popupFactory?: PopupFactory;


}

/**
 * A component for displaying a Leaflet map and markers for popups for locations where weather data is being tracked.
 * 
 * @component
 * 
 */

function LeafletMap({ markerCallback, mapCentre, marker, popupFactory }: MapProps): JSX.Element {

  const [popupComponent, setPopupComponent] = useState<React.ReactNode>();
  const [dragging, setDragging] = useState(false);
  const [display, setDisplay] = useState(false);

  function closePopup() {

    setPopupComponent(null);
    setDragging(true);
    setDisplay(false);

  }

  useEffect(() => {
    if (display && popupFactory) {

      const popupComponent = popupFactory(closePopup);

      setPopupComponent(popupComponent);
      setDragging(false);
    }
  }, [popupFactory, display]);


  function mCallback(markerName: string) {

    setDisplay(true);
    markerCallback(markerName);

  }

  return (
    <Map center={mapCentre} zoom={DEFAULT_ZOOM} dragging={dragging} >

      <TileLayer
        attribution='&copy; <a href="https://osm.org/copyright">OpenStreetMap</a> contributors'
        url='https://{s}.tile.osm.org/{z}/{x}/{y}.png'
      />

      {marker.map((createMarker) => createMarker(this, mCallback, image))}

      <div id="marginclickdiv" >
        {popupComponent}
      </div>
    </Map>
  );

}

LeafletMap.propTypes = {


  /** An array of markers to be displayed on the map. Each element of the array specifies the coordinates and location 
   * name associated with a marker.*/
  marker: PropTypes.array,

  /** A function called when a marker is clicked. */
  markerCallback: PropTypes.func,

  /** The latitude/longitude coordinates for the centre of the map. */
  mapCentre: PropTypes.array.isRequired,


  popupFactory: PropTypes.func,
}

export default LeafletMap;
