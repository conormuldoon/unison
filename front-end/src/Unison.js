

import PropTypes from 'prop-types';
import React, { useState, useEffect } from 'react';
import 'react-day-picker/lib/style.css';
import { formatDate } from 'react-day-picker/moment';
import './App.css';
import ARLocationComponent from './ARLocationComponent';
import { FORMAT, SELF } from './Constant';
import createSelector from './selectorFactory';
import { today, tomorrow, expandLink, problemConnecting } from './Util';
import HttpStatus from 'http-status-codes';

import { createLocationFactory, createRemoveFactory } from './closureFactory';

/**
 * Application component for Unison. Once mounted, it connects to the back-end to receive a list of the locations being tracked.
 * 
 * @component
 * 
 */
function Unison(props) {


  const [fromDate, setFromDate] = useState(today());
  const [toDate, setToDate] = useState(tomorrow());
  const [option, setOption] = useState(null);
  const [marker, setMarker] = useState(null);
  const [curLoc, setCurLoc] = useState(null);
  const [curVar, setCurVar] = useState(null);
  const [varOpt, setVarOpt] = useState(null);
  const [locationMap, setLocationMap] = useState(null);
  const [collectionModel, setCollectionModel] = useState(null);

  const obtainData = () => {

    class ResponseError extends Error {
      constructor(message, status) {
        super(message);
        this.status = status;
      }
    }

    function checkResponse(response, endpoint) {
      if (!response.ok)
        throw new ResponseError('Bad response for ' + endpoint, response.status);
    }

    async function requestFeatureCollection(uri) {
      const response = await fetch(uri, {
        method: 'GET',
        headers: new Headers({
          'Accept': 'application/geo+json'
        })
      });


      checkResponse(response, uri);

      const fc = await response.json();

      const locationArray = fc.features;
      const n = locationArray.length;

      const newOption = [];
      const newMarker = [];

      for (let i = 0; i < n; i++) {
        newOption.push(locationArray[i].properties.name);
        let pos = [locationArray[i].geometry.coordinates[1], locationArray[i].geometry.coordinates[0]];

        const properties = locationArray[i].properties;
        newMarker.push({ name: properties.name, position: pos });

      }

      if (n > 0) {

        setOption(newOption);
        setMarker(newMarker);


      } else {
        setOption(null);
        setMarker(null);
      }


    }

    const halGetHeader = {
      method: 'GET',
      headers: new Headers({
        'Accept': 'application/hal+json'
      }),

    }

    function clearCurrent() {
      setCurLoc(null);
      setCurVar(null);
      setVarOpt(null);
      setLocationMap(null);

    }

    async function requestCollectionHAL(uri) {

      const response = await fetch(uri, halGetHeader);


      checkResponse(response, uri);

      const model = await response.json();


      setCollectionModel(model);

      if (model._embedded) {
        const list = model._embedded.locationModelList;
        const n = list.length;
        const map = new Map();
        for (let i = 0; i < n; i++) {
          map.set(list[i].name, list[i]);
        }

        setLocationMap(map);

        const varOpt = [];

        if (n > 0) {

          for (const rel in list[0]._links) {
            if (SELF === rel) {
              continue;
            }

            const s = rel.charAt(0).toUpperCase() + rel.substring(1);
            let optName = '';
            let sw = 0;
            for (let i = 1; i < s.length; i++) {
              const c = s.charAt(i);
              if (c >= 'A' && c <= 'Z') {
                optName += s.substring(sw, i) + ' ';
                sw = i;
              }
            }
            optName += s.substring(sw);
            varOpt.push(optName);
          }
        }
        if (varOpt.length > 0) {
          setCurLoc(list[0]);
          setCurVar(varOpt[0]);
          setVarOpt(varOpt);
          return;

        } else {

          clearCurrent();
          alert("No weather data is associated with the collection.")
        }


      } else {
        clearCurrent();
      }


    }

    async function requestModel() {
      try {

        const root = process.env.PUBLIC_URL;

        const response = await fetch(root, halGetHeader);

        checkResponse(response, root);


        const model = await response.json();

        const uri = model._links.locationCollection.href;
        await requestCollectionHAL(uri);
        await requestFeatureCollection(uri);

      } catch (e) {
        if (e instanceof ResponseError) {
          console.error("HTTP status code: " + e.status);

        }
        problemConnecting();

      }

    }


    requestModel();




  };

  useEffect(() => {

    obtainData();
  }, []);


  const updateCurLoc = (locationName) => {
    setCurLoc(locationMap.get(locationName));
  }

  const markerClicked = (location) => {
    updateCurLoc(location);
  }

  const _onLocationSelect = (event) => {

    updateCurLoc(event.target.value);

  }

  const _onVarSelect = (event) => {

    setCurVar(event.target.value);

  }

  const handleStartChange = (selectedDate) => {
    setFromDate(formatDate(selectedDate, FORMAT));

  }

  const handleEndChange = (selectedDate) => {

    setToDate(formatDate(selectedDate, FORMAT));

  }

  const spaceToUnderscore = (s) => {
    return s.replace(/ /g, '_')
  }


  const handleCSV = async () => {

    const response = await fetch(expandLink(curLoc, curVar, fromDate, toDate));

    if (response.ok) {
      const blob = await response.blob();
      const a = document.createElement('a');
      a.href = window.URL.createObjectURL(blob);
      a.download = spaceToUnderscore(curLoc.name) + '_' + spaceToUnderscore(curVar) + '_'
        + fromDate + "_" + toDate + '.csv';
      a.click();
    } else if (response.status === HttpStatus.GATEWAY_TIMEOUT) {
      problemConnecting();
    }
  }

  return (

    <div id="mapdiv">

      <div id="logos">

        <center>
          {props.logoLeft}
          {props.logoRight}
        </center>

      </div>

      {props.createMap(marker, curVar, locationMap, markerClicked, fromDate, toDate, curLoc)}

      <div id="selectdiv" >

        <center>

          <div>
            {curLoc && <div id='variDD' >

              {varOpt && <select onChange={_onVarSelect}>

                {varOpt.map((opt) => <option key={opt} value={opt}>{opt}</option>)}

              </select>}

            </div>}

            {curLoc && <div className='marginItem' >

              <select onChange={_onLocationSelect} value={curLoc.name}>
                {!option && <option key="Location" value="Location" >Location</option>}
                {option && option.map((opt) => <option key={opt} value={opt}>{opt}</option>)}

              </select>

            </div>}


            <div className='pLeft' >{
              createSelector("From date", fromDate, handleStartChange)
            }
            </div>

            <div className='pLeft' >
              {
                createSelector("To date", toDate, handleEndChange)
              }
            </div>

            {curLoc && curVar && <div className='pLeft'> <button onClick={handleCSV}>CSV</button> </div>}
          </div>

          <ARLocationComponent createLocation={createLocationFactory(obtainData, collectionModel)}
            createRemove={curLoc ? createRemoveFactory(obtainData, curLoc._links[SELF].href, curLoc.name) : null} />

        </center>

      </div>



    </div>
  );

}

Unison.propTypes = {

  /** A function for creating the map. */
  createMap: PropTypes.func.isRequired,

  /** A logo displayed at the bottom of the screen. It will be displayed to the left
   * if the logoRight prop is defined.
   */
  logoLeft: PropTypes.any,

  /** A logo displayed at the bottom of the screen. It will be displayed to the right
   * if the logoLeft prop is defined.
   */
  logoRight: PropTypes.any,

}

export default Unison;
