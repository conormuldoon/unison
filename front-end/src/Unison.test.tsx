
import { enableFetchMocks } from 'jest-fetch-mock';
enableFetchMocks();
import "@testing-library/jest-dom/extend-expect";
import fetchMock from 'jest-fetch-mock';
import React from 'react';
import { fireEvent, render, screen, waitForElementToBeRemoved } from "@testing-library/react";
import Unison from './Unison';
import { createMapFactory } from './LeafletMap';




const unisonModel = {
    "_links": {
        "self": [
            {
                "href": "http://localhost:8080/"
            },
            {
                "href": "http://localhost:8080/index"
            }
        ],
        "locationCollection": {
            "href": "http://localhost:8080/locationCollection"
        },
        "user": {
            "href": "http://localhost:8080/user"
        }
    }
};

const locationModelHAL = {
    "_embedded": {
        "locationModelList": [
            {
                "name": "UCD",
                "_links": {
                    "self": {
                        "href": "http://localhost:8080/locationCollection/UCD"
                    },
                    "cloudiness": {
                        "href": "http://localhost:8080/locationCollection/UCD/cloudiness{?fromDate,toDate}",
                        "templated": true
                    },
                    "cloudLevel": {
                        "href": "http://localhost:8080/locationCollection/UCD/cloudLevel{?fromDate,toDate}",
                        "templated": true
                    },
                    "dewPoint": {
                        "href": "http://localhost:8080/locationCollection/UCD/dewPoint{?fromDate,toDate}",
                        "templated": true
                    },
                    "humidity": {
                        "href": "http://localhost:8080/locationCollection/UCD/humidity{?fromDate,toDate}",
                        "templated": true
                    },
                    "precipitation": {
                        "href": "http://localhost:8080/locationCollection/UCD/precipitation{?fromDate,toDate}",
                        "templated": true
                    },
                    "pressure": {
                        "href": "http://localhost:8080/locationCollection/UCD/pressure{?fromDate,toDate}",
                        "templated": true
                    },
                    "temperature": {
                        "href": "http://localhost:8080/locationCollection/UCD/temperature{?fromDate,toDate}",
                        "templated": true
                    },
                    "windDirection": {
                        "href": "http://localhost:8080/locationCollection/UCD/windDirection{?fromDate,toDate}",
                        "templated": true
                    },
                    "windSpeed": {
                        "href": "http://localhost:8080/locationCollection/UCD/windSpeed{?fromDate,toDate}",
                        "templated": true
                    }
                }
            }
        ]
    },
    "_links": {
        "self": {
            "href": "http://localhost:8080/locationCollection"
        },
        "contains": {
            "href": "http://localhost:8080/locationCollection/contains{?name}",
            "templated": true
        }
    }
};

const locationCollection = {
    "type": "FeatureCollection",
    "features": [
        {
            "type": "Feature",
            "geometry": {
                "type": "Point",
                "coordinates": [
                    -6.4,
                    53.4
                ]
            },
            "properties": {
                "name": "UCD"
            }
        }
    ]
};


const mapCentre: [number, number] = [59.922326, 10.751560];
const mapFactory = createMapFactory(mapCentre);


describe('Unison', () => {

    beforeEach(() => {
        fetchMock.mockResponses(
            JSON.stringify(unisonModel),
            JSON.stringify(locationModelHAL),
            JSON.stringify(locationCollection));
    });

    it('renders without crashing', async () => {


        render(<Unison createMap={mapFactory} />);
        await screen.findByRole('button', { name: 'CSV' });



    });

    it('mathes snapshot', async () => {


        const mockDateNow = jest.fn(() => 1571875200000);
        const dn = global.Date.now;
        global.Date.now = mockDateNow;

        try {
            const { container } = render(<Unison createMap={mapFactory} />);
            await screen.findByRole('button', { name: 'CSV' });
            expect(container).toMatchSnapshot();
        } finally {
            global.Date.now = dn;
        }


    });

    it('displays a button to remove the first location received', async () => {
        render(<Unison createMap={mapFactory} />);
        await screen.findByRole('button', {
            name: 'Remove ' +
                locationCollection.features[0].properties.name
        });
    });

    it('does not dispplay a remove button when no loctioos', async () => {
        render(<Unison createMap={mapFactory} />);
        const removeButton = await screen.findByRole('button', {
            name: 'Remove ' +
                locationCollection.features[0].properties.name
        });

        const emptyFC = {
            "type": "FeatureCollection",
            "features": []
        }

        const emptyHALC = {

            "_links": {
                "self": {
                    "href": "http://localhost:8080/locationCollection"
                },
                "contains": {
                    "href": "http://localhost:8080/locationCollection/contains{?name}",
                    "templated": true
                }
            }
        }

        fetchMock.mockResponses(
            "",
            JSON.stringify(unisonModel),
            JSON.stringify(emptyHALC),
            JSON.stringify(emptyFC)
        );

        const alertSpy = jest.spyOn(window, 'alert');
        alertSpy.mockImplementation(() => { });

        const confirmSpy = jest.spyOn(window, 'confirm');
        confirmSpy.mockImplementation(() => true);

        try {
            fireEvent.click(removeButton);

            

            await waitForElementToBeRemoved(removeButton);

        } finally {
            confirmSpy.mockClear();
            alertSpy.mockClear();
        }

    });

    it('displays popup when marker clicked', async () => {




        const precipData = [{
            "date": "2019-04-01T23:00:00.000+0000",
            "precipitation": {
                "value": 0,
                "minvalue": 0,
                "maxvalue": 0.1
            }
        },
        {
            "date": "2019-04-02T00:00:00.000+0000",
            "precipitation": {
                "value": 0,
                "minvalue": 0,
                "maxvalue": 0
            }
        },
        {
            "date": "2019-04-02T01:00:00.000+0000",
            "precipitation": {
                "value": 0,
                "minvalue": 0,
                "maxvalue": 0
            }
        },
        {
            "date": "2019-04-02T02:00:00.000+0000",
            "precipitation": {
                "value": 0,
                "minvalue": 0,
                "maxvalue": 0
            }
        },
        {
            "date": "2019-04-02T03:00:00.000+0000",
            "precipitation": {
                "value": 0,
                "minvalue": 0,
                "maxvalue": 0
            }
        },
        {
            "date": "2019-04-02T04:00:00.000+0000",
            "precipitation": {
                "value": 0,
                "minvalue": 0,
                "maxvalue": 0
            }
        },
        {
            "date": "2019-04-02T05:00:00.000+0000",
            "precipitation": {
                "value": 0,
                "minvalue": 0,
                "maxvalue": 0
            }
        },
        {
            "date": "2019-04-02T06:00:00.000+0000",
            "precipitation": {
                "value": 0,
                "minvalue": 0,
                "maxvalue": 0
            }
        },
        {
            "date": "2019-04-02T07:00:00.000+0000",
            "precipitation": {
                "value": 0,
                "minvalue": 0,
                "maxvalue": 0
            }
        },
        {
            "date": "2019-04-02T08:00:00.000+0000",
            "precipitation": {
                "value": 0,
                "minvalue": 0,
                "maxvalue": 0
            }
        },
        {
            "date": "2019-04-02T09:00:00.000+0000",
            "precipitation": {
                "value": 0,
                "minvalue": 0,
                "maxvalue": 0
            }
        },
        {
            "date": "2019-04-02T10:00:00.000+0000",
            "precipitation": {
                "value": 0,
                "minvalue": 0,
                "maxvalue": 0
            }
        }];


        render(<Unison createMap={mapFactory} />);

        fetchMock.mockResponses(
            JSON.stringify(precipData),
            JSON.stringify({ value: false })
        );

        await screen.findAllByAltText('');

        fireEvent.click(screen.getAllByAltText('')[1]);


        await screen.findByText('UCD');



    });
});


