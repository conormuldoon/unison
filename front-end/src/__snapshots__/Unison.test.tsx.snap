// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unison mathes snapshot 1`] = `
<div>
  <div
    id="mapdiv"
  >
    <div
      id="logos"
      style="text-align: center;"
    />
    <div
      class="leaflet-container leaflet-grab leaflet-touch-drag"
      style="position: relative;"
      tabindex="0"
    >
      <div
        class="leaflet-pane leaflet-map-pane"
        style="left: 0px; top: 0px;"
      >
        <div
          class="leaflet-pane leaflet-tile-pane"
        >
          <div
            class="leaflet-layer "
            style="z-index: 1;"
          >
            <div
              class="leaflet-tile-container leaflet-zoom-animated"
              style="z-index: 18; left: 0px; top: 0px;"
            >
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="https://b.tile.osm.org/12/2170/1191.png"
                style="width: 256px; height: 256px; left: -84px; top: -62px;"
              />
            </div>
          </div>
        </div>
        <div
          class="leaflet-pane leaflet-shadow-pane"
        />
        <div
          class="leaflet-pane leaflet-overlay-pane"
        />
        <div
          class="leaflet-pane leaflet-marker-pane"
        >
          <img
            alt=""
            class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
            src="2000px-Map_marker.png"
            style="margin-left: -15px; margin-top: -23px; width: 30px; height: 46px; left: -49957px; top: 34673px; z-index: 34673;"
            tabindex="0"
          />
        </div>
        <div
          class="leaflet-pane leaflet-tooltip-pane"
        />
        <div
          class="leaflet-pane leaflet-popup-pane"
        />
      </div>
      <div
        class="leaflet-control-container"
      >
        <div
          class="leaflet-top leaflet-left"
        >
          <div
            class="leaflet-control-zoom leaflet-bar leaflet-control"
          >
            <a
              aria-label="Zoom in"
              class="leaflet-control-zoom-in"
              href="#"
              role="button"
              title="Zoom in"
            >
              +
            </a>
            <a
              aria-label="Zoom out"
              class="leaflet-control-zoom-out"
              href="#"
              role="button"
              title="Zoom out"
            >
              −
            </a>
          </div>
        </div>
        <div
          class="leaflet-top leaflet-right"
        />
        <div
          class="leaflet-bottom leaflet-left"
        />
        <div
          class="leaflet-bottom leaflet-right"
        >
          <div
            class="leaflet-control-attribution leaflet-control"
          >
            <a
              href="https://leafletjs.com"
              title="A JS library for interactive maps"
            >
              Leaflet
            </a>
             | © 
            <a
              href="https://osm.org/copyright"
            >
              OpenStreetMap
            </a>
             contributors
          </div>
        </div>
      </div>
      <div
        id="marginclickdiv"
      />
    </div>
    <div
      id="selectdiv"
      style="text-align: center;"
    >
      <div>
        <div
          id="variDD"
        >
          <select>
            <option
              value="Cloudiness"
            >
              Cloudiness
            </option>
            <option
              value="Cloud Level"
            >
              Cloud Level
            </option>
            <option
              value="Dew Point"
            >
              Dew Point
            </option>
            <option
              value="Humidity"
            >
              Humidity
            </option>
            <option
              value="Precipitation"
            >
              Precipitation
            </option>
            <option
              value="Pressure"
            >
              Pressure
            </option>
            <option
              value="Temperature"
            >
              Temperature
            </option>
            <option
              value="Wind Direction"
            >
              Wind Direction
            </option>
            <option
              value="Wind Speed"
            >
              Wind Speed
            </option>
          </select>
        </div>
        <div
          class="marginItem"
        >
          <select>
            <option
              value="UCD"
            >
              UCD
            </option>
          </select>
        </div>
        <div
          class="pLeft"
        >
          <div>
            <p>
              From date
            </p>
            <div
              class="DayPickerInput"
            >
              <input
                placeholder="YYYY-M-D"
                value="24-10-2019"
              />
            </div>
          </div>
        </div>
        <div
          class="pLeft"
        >
          <div>
            <p>
              To date
            </p>
            <div
              class="DayPickerInput"
            >
              <input
                placeholder="YYYY-M-D"
                value="25-10-2019"
              />
            </div>
          </div>
        </div>
        <div
          class="pLeft"
        >
           
          <button>
            CSV
          </button>
           
        </div>
      </div>
      <div>
        <div
          class="pLeft"
        >
          <div
            id="pLeft"
          >
            <button
              data-testid="lf-button"
            >
              Add Location
            </button>
          </div>
        </div>
        <div
          class="pLeft"
        >
          <button
            data-testid="rm-button"
          >
            Remove UCD
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
